prov/rxm/src/rxm_rma.c

```diff
@@ -52,17 +52,253 @@ rxm_ep_rma_reg_iov(struct rxm_ep *rxm_ep, const struct iovec *msg_iov,
+	msg_rma->desc = desc;
+	msg_rma->iov_count = iov_count;
+	msg_rma->addr = orig_msg->addr;
+	msg_rma->rma_iov = orig_msg->rma_iov;
+	msg_rma->rma_iov_count = orig_msg->rma_iov_count;
+	msg_rma->context = tx_entry;
+	msg_rma->data = orig_msg->data;
+}
+
+static inline void
+rxm_ep_rma_fill_msg_no_buf(struct rxm_rma_buf *rma_buf,
+			   struct rxm_tx_entry *tx_entry,
+			   const struct fi_msg_rma *orig_msg)
+{
+    uint8_t i;
+
+	rma_buf->rxm_iov.count = (uint8_t)orig_msg->iov_count;
+	for (i = 0; i < rma_buf->rxm_iov.count; i++) {
+		rma_buf->rxm_iov.iov[i].iov_base = orig_msg->msg_iov[i].iov_base;
+		rma_buf->rxm_iov.iov[i].iov_len = orig_msg->msg_iov[i].iov_len;
+	}
+
+	rxm_ep_rma_fill_msg(&rma_buf->msg, rma_buf->rxm_iov.iov,
+			    rma_buf->rxm_iov.count, rma_buf->rxm_iov.desc,
+			    tx_entry, orig_msg);
+}
+
+static inline void
+rxm_ep_rma_fill_msg_buf(struct rxm_rma_buf *rma_buf,
+			struct rxm_tx_entry *tx_entry,
+			const struct fi_msg_rma *orig_msg)
+{
+	ofi_copy_from_iov(rma_buf->pkt.data, rma_buf->pkt.hdr.size,
+			  orig_msg->msg_iov, orig_msg->iov_count, 0);
+
+	rma_buf->rxm_iov.iov[0].iov_base = &rma_buf->pkt.data;
+	rma_buf->rxm_iov.iov[0].iov_len = rma_buf->pkt.hdr.size;
+
+	rxm_ep_rma_fill_msg(&rma_buf->msg, rma_buf->rxm_iov.iov,
+			    1, &rma_buf->hdr.desc, tx_entry, orig_msg);
+}
+
+static inline ssize_t
+rxm_ep_format_rma_res_lightweight(struct rxm_ep *rxm_ep, uint64_t flags,
+				  uint64_t comp_flags, const struct fi_msg_rma *orig_msg,
+				  struct rxm_tx_entry **tx_entry)
+{
+	*tx_entry = rxm_tx_entry_get(&rxm_ep->send_queue);
+	if (OFI_UNLIKELY(!tx_entry))
+		return -FI_EAGAIN;
+
+	(*tx_entry)->state = RXM_TX_NOBUF;
+	(*tx_entry)->context = orig_msg->context;
+	(*tx_entry)->flags = flags;
+	(*tx_entry)->comp_flags = FI_RMA | comp_flags;
+	(*tx_entry)->count = orig_msg->iov_count;
+
+	return FI_SUCCESS;
+}
+
+static inline ssize_t
+rxm_ep_format_rma_buf(struct rxm_ep *rxm_ep, size_t total_size,
+		      struct rxm_rma_buf **rma_buf, struct rxm_tx_entry *tx_entry)
+{
+	*rma_buf = rxm_rma_buf_get(rxm_ep);
+	if (OFI_UNLIKELY(!*rma_buf))
+		return -FI_EAGAIN;
+
+	tx_entry->state = RXM_TX_RMA;
+	tx_entry->rma_buf = *rma_buf;
+	(*rma_buf)->pkt.hdr.size = total_size;
+
+	return FI_SUCCESS;
+}
+
+static inline ssize_t
+rxm_ep_format_rma_res(struct rxm_ep *rxm_ep, size_t total_size,
+		      uint64_t flags, uint64_t comp_flags,
+		      const struct fi_msg_rma *orig_msg,
+		      struct rxm_rma_buf **rma_buf,
+		      struct rxm_tx_entry **tx_entry)
+{
+	ssize_t ret;
+
+	ret = rxm_ep_format_rma_res_lightweight(rxm_ep, flags, comp_flags,
+						orig_msg, tx_entry);
+	if (OFI_UNLIKELY(ret))
+		goto err1;
+
+	ret = rxm_ep_format_rma_buf(rxm_ep, total_size,
+				    rma_buf, *tx_entry);
+	if (OFI_UNLIKELY(!ret))
+		goto err2;
+
+	return FI_SUCCESS;
+err2:
+	rxm_tx_entry_release(&rxm_ep->send_queue, *tx_entry);
+err1:
+	FI_WARN(&rxm_prov, FI_LOG_CQ, "Unable to allocate resources!\n");
+	rxm_ep_progress_multi(&rxm_ep->util_ep);
```

Comment:
This looks like a side effect for a function called 'format_rma_res'.

Reply:
Sorry, I'm not sure that I got it correct. What do you mean? Do you mean that `err1` label is excessive?

Reply:
The name of the function suggests that all this call does is to format some sort of request.  Calling 'progress' in the case of a failure comes across as an unexpected operation.  I didn't see this in the other 'format' calls.

Reply:
Oh, I got it. Thanks. Removed this call